### PR TITLE
feat: script to annotate proteome hits onto winnow outputs

### PR DESCRIPTION
**Pull request by JemmaLDaniel**
## Summary

Adds a post-`winnow predict` utility that annotates prediction outputs with a `proteome_hit` column by matching peptides against a reference FASTA proteome.

The command drops short peptides (default: fewer than 7 tokeniser residues), strips modifications and normalises I/L for matching, then uses Aho-Corasick substring search. It updates `preds_and_fdr_metrics.csv` in place and filters `metadata.csv` to the retained spectra.

The resulting `proteome_hit` column is designed to feed `winnow diagnose-calibration` via `diagnostics.label_source=precomputed diagnostics.label_column=proteome_hit`.

## Changes

- New library module `winnow/scripts/annotate_proteome_hits.py` holding the matching logic (FASTA loading, modification stripping, residue tokenisation, batched substring search, folder annotation).
- New CLI subcommand `winnow annotate-proteome-hits` on the main Typer app in `winnow/scripts/main.py`, following the existing lazy-import pattern..
- Tokenisation uses `instanovo.utils.residues.ResidueSet.tokenize`.
- `biopython` and `pyahocorasick` are added as an optional `proteome` extra (`pip install 'winnow-fdr[proteome]'`), not as core dependencies. Their imports are lazy and guarded, so the base install and `winnow.scripts.main` remain importable without the extra.
- Documentation for the new command and the `[proteome]` extra added to `docs/cli.md`.
- Unit tests in `tests/scripts/test_annotate_proteome_hits.py` covering modification stripping, short-peptide filtering, the cross-protein separator guarantee, residue token counting edge cases, and a full folder round-trip.

### CODE REVIEW

File: winnow/utils/proteome.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+import ahocorasick
+import polars as pl
+import yaml
+from Bio import SeqIO
+from instanovo.utils.residues import ResidueSet
+
+logger = logging.getLogger(__name__)
+
+_PROTEOME_JOIN_SEP = "\x1f"
```
Comment (BioGeek) [inline winnow/utils/proteome.py:22]: Cool, learned something new today

> The notation `\x1f` represents the ASCII unit separator, a non-printable control character with the hexadecimal value 1f (decimal 31). It is commonly used in programming and data streams as an invisible delimiter to separate distinct units or fields of data within a single string.

File: winnow/utils/proteome.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        if sequence:
+            sequences.append(sequence)
+    return _PROTEOME_JOIN_SEP.join(sequences)
+
+
+def processed_peptide_for_match(prediction: str) -> str:
+    """Strip mods and normalise I/L for proteome substring matching."""
+    if not prediction or not isinstance(prediction, str):
+        return ""
+    sequence = _MOD_ROUND.sub("", prediction)
```
Comment (BioGeek) [inline winnow/utils/proteome.py:69]: `ResidueSet` can also tokenize raw numeric modification notation such as `M+15.99PEPTIDE` but these aren't handled here and would leave  `+15.99` in the automaton pattern. So strip the tokenizer's raw numeric modification forms as well, including terminal forms.

File: winnow/scripts/annotate_proteome_hits.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        preds, haystack, residue_set, min_residue_length=min_residue_length
+    )
+    num_kept = annotated.height
+    num_removed_short = num_input - num_kept
+
+    kept_spectrum_ids = annotated.select("spectrum_id").unique()
+    metadata = pl.read_csv(meta_path)
+    if "spectrum_id" not in metadata.columns:
+        raise ValueError(f"'spectrum_id' column missing in {meta_path}")
+    metadata_kept = metadata.join(kept_spectrum_ids, on="spectrum_id", how="inner")
```
Comment (BioGeek) [inline winnow/scripts/annotate_proteome_hits.py]: Build a sample model and predict output exactly as defined in the Makefile.
```
make train-sample
make predict-sample
```

`results/predictions/` now contains the two files produced by winnow predict.

Supply a tiny reference proteome containing one known sample prediction after modification stripping.

 ```
printf '>sample_protein\nXXHQGVMVGMGQKXX\n' >/tmp/winnow-proteome.fasta
 ```

Annotate the sample predict output.

 ```
uv run winnow annotate-proteome-hits results/predictions --fasta /tmp/winnow-proteome.fasta
 ```
Verify where the new label was written:

 ```python
import pandas as pd

folder = "results/predictions"
preds = pd.read_csv(f"{folder}/preds_and_fdr_metrics.csv")
metadata = pd.read_csv(f"{folder}/metadata.csv")

print("preds has proteome_hit:", "proteome_hit" in preds.columns)
print("metadata has proteome_hit:", "proteome_hit" in metadata.columns)
print("metadata has prediction:", "prediction" in metadata.columns)
print("matched sample hit:", preds["proteome_hit"].any())
 ```

Current output:

 ```
preds has proteome_hit: True
metadata has proteome_hit: False
metadata has prediction: False
matched sample hit: True
```

Finally, run the diagnostic workflow against the annotated folder:

```
uv run winnow diagnose-calibration \
    calibrator.pretrained_model_name_or_path=models/new_model \
    data_loader=winnow \
    dataset.spectrum_path_or_directory=results/predictions \
    dataset.predictions_path=null \
    diagnostics.label_source=precomputed \
    diagnostics.label_column=proteome_hit
 ```

It fails with a `KeyError: 'prediction'` because the Winnow loader reads only `results/predictions/metadata.csv`, which lacks prediction (and also lacks `proteome_hit`). The annotation command wrote the label exclusively to `preds_and_fdr_metrics.csv`, making the diagnostic workflow unusable. So join the annotation columns into metadata_kept before writing it.

**Review by BioGeek — CHANGES_REQUESTED** [review]

File: winnow/utils/proteome.py (hunk 2 above)
Comment (JemmaLDaniel) [reply winnow/utils/proteome.py:69]: Addressed in [94b5343](https://github.com/instadeepai/winnow/pull/220/commits/94b53433b5bc893bd3f8b3fd63f7ba46a473b6c3)

File: winnow/scripts/annotate_proteome_hits.py (hunk 3 above)
Comment (JemmaLDaniel) [reply winnow/scripts/annotate_proteome_hits.py]: Thanks for this, I rethought the design and have changed it to instead match the existing workflow of 1) load with a data loader and standardise to CalibrationDataset, 2) perform operations, 3) save as a `CalibrationDataset`. This way we avoid the problem you found above by annotating a dataset from any supported data source, saving as a Winnow dataset with `CalibrationDataset`'s `save()`, and then reloading for calibration diagnostics by setting the data loader to `winnow`. I also think it fits into the existing codebase more cleanly. The downside is that the annotation operation is now less lightweight, with all the extra verification and standardisation steps.

**Review by BioGeek — APPROVED** [review]
The refactoring lookd good, thanks!
